python/google/protobuf/pyext/descriptor_pool.cc

```diff
@@ -121,24 +215,43 @@ static PyDescriptorPool* _CreateDescriptorPool() {
     return NULL;
   }
 
+  cpool->in_process_database = nullptr;
+  cpool->file_error_collector = nullptr;
+  cpool->disk_source_tree = nullptr;
+  cpool->disk_database = nullptr;
+
   PyObject_GC_Track(cpool);
 
   return cpool;
 }
 
-// Create a Python DescriptorPool, using the given pool as an underlay:
-// new messages will be added to a custom pool, not to the underlay.
-//
-// Ownership of the underlay is not transferred, its pointer should
-// stay alive.
-static PyDescriptorPool* PyDescriptorPool_NewWithUnderlay(
-    const DescriptorPool* underlay) {
+/* Initializes a PyDescriptorPool like the default descriptor pool.
+ *
+ * This method returns a PyDescriptorPool able to cross-link protos
+ * originating from the C++ generated pool, from Python space as
+ * FileDescriptorProtos in memory (like for generated pb2 files) or from the
+ * filesystem as .proto files.
+ *
+ * This method adds a database to enable the latter two cases and an underlay
+ * DescriptorPool to handle the former case. BuildFile must not be called on the
+ * a this object's underlying DescriptorPool object. Instead, objects should be
+ * registered into its in-process database using the Register method. Then,
+ * FileDescriptor objects may be obtained by calling FindFileByName.
+ */
+static PyDescriptorPool* PyDescriptorPool_NewDefault() {
   PyDescriptorPool* cpool = _CreateDescriptorPool();
   if (cpool == NULL) {
     return NULL;
   }
-  cpool->pool = new DescriptorPool(underlay);
-  cpool->underlay = underlay;
+  cpool->underlay = DescriptorPool::generated_pool();
+  cpool->file_error_collector = new PyErrorCollector;
+  cpool->disk_source_tree = new DiskSourceTree;
+  cpool->disk_database = new SourceTreeDescriptorDatabase(cpool->disk_source_tree);
```

Comment:
Always create disk database to the generated pool will change current  pool behavior I think? For a project that generated some .proto file to _pb2 and not generated some other .proto file, descriptor pool can only find those are generated before this change but can find all  the files after this PR

Reply:
As far as I can tell, this does not actually change the behavior of the current pool. `InProcessDescriptorDatabase` gives precedence to descriptors added from generated files, so the `SourceTreeDescriptorDatabase` will only ever be used for descriptors that cannot be found in a generated `_pb2.py` file.

Importing a module is an opt-in process. So only applications that are currently expecting an import to fail should be surprised by this change.

This PR is currently passing the suite of unit tests. If any of this PR constitutes a breaking change, shouldn't we add tests to capture said behavior?

Reply:
"applications that are currently expecting an import to fail should be surprised by this change" This is the behavior change I was talking about. The issues for the behavior change:
1, Pure python and cpp extension will have different result when find file in pool
2, For users that did not use the protobuf.protos(), they will able to find the descriptors from pool but not able to use the modules. Which is confusing.

Reply:
1. Can you expand on your first point? Is the expectation that the pure Python implementation will behave exactly like the C extension?
2. What would you suggest instead? Initialize the `InProcessDescriptorPool` so that it doesn't use a fallback `SourceTreeDescriptorDatabase` and only enable it the first time the user calls `protobuf.protos()`?

Reply:
1, Not exactly same. But we are always trying to fix the behavior differences (there are some differences unable to be fixed though). We do not want to introduce more differences.

2, How about not use SourceTreeDescriptorDatabase as fallback for the default pool. Instead: Create a static SourceTreeDescriptorDatabase to parse files when protobuf.protos() is called. And then add the parsed FileDescriptor to the default pool.

Reply:
I explored that option, but I couldn't get it work robustly. In order to build `FileDescriptor` objects, you need a `DescriptorPool` for them to live in. So there are two options for which pool to use:
 1. Use a separate pool from the default Python-level pool.
 2. Directly use the default Python-level pool. (essentially the current state of this PR)

If you use a separate pool, there are a few problems. Suppose `B.proto` depends on `A.proto`. The application first does `import A_pb2`. Then it does `B = protobuf.protos("B.proto")`. `B.proto` needs to be linked with the `A.proto` `FileDescriptor`, which lives in the default Python `DescriptorPool`. You could solve this by using a fallback db pointing at the default Python `DescriptorPool`, but then the opposite dependency direction becomes a problem. That is, the following code would no longer work:
```python
A = protobuf.protos("A.proto")
import B_pb2
```

Or, at the very least, it would create a second copy of `A.proto`'s module loaded from an `A_pb2.py` which could lead to pointer equality problems.

By putting both statically generated and runtime-parsed protos into the same pool, both scenarios above work out, which is explained in detail in the docstring for `protobuf.protos()`.

If the concern is that we don't want to change the behavior for users who don't opt in by using the new function, shouldn't it be sufficient to only start using the `SourceTreeDescriptorDatabase` once `protobuf.protos()` has been called for the first time?

Reply:
Would add an internal_set_fallback_database() API to descriptor pool work? Thus we can turn on SourceTreeDescriptorDatabase only in protobuf.protos(), and turn it off before exit protobuf.protos()

Reply:
@anandolee I think that would work if `internal_set_fallback_database` is a method on the new `InProcessDescriptorDatabase`. The whole reason that it exists (instead of just using the method you're suggesting) is because [you can't call `BuildFile` on a `DescriptorPool` with a fallback database set](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.cc#L3559). Other than that, the finer grained control over the usage of the `SourceTreeDescriptorDatabase` sounds like a good idea to me.

Reply:
@anandolee Friendly ping.

Reply:
" you can't call BuildFile on a DescriptorPool with a fallback database set. "  no need call BuildFile on descriptor pool, only FileFileByName() will be called and then invoke DescriptorPool::TryFindFileInFallbackDatabase().   But add internal_set_fallback_database() in InProcessDescriptorDatabase sounds better to me, because we do not need to change protobuf c++ APIs, the change will be in python c extension only. I'd still like to suggest wait for David's review

Reply:
While I appreciate that the DescriptorPool behavior demonstrates the issue, I think this is the same problem with the importer.

This facility already shouldn't be touching the import machinery, which also means it shouldn't be touching the global DescriptorPool.

Reply:
@anandolee 

> no need call BuildFile on descriptor pool

This PR may not be calling `BuildFile`, but [the default Python-level descriptor pool uses that method to register the descriptors from the generated files](https://github.com/protocolbuffers/protobuf/blob/4cf5bfee9546101d98754d23ff378ff718ba8438/python/google/protobuf/pyext/descriptor_pool.cc#L608). This is the operation that would fail if we set a fallback db and an underlay pool on the default Python pool.

@dlj-NaN

> This facility already shouldn't be touching the import machinery, which also means it shouldn't be touching the global DescriptorPool.

I can see the argument to be made here. We need the set of ".proto" files loadable as modules to mirror one-to-one the set of FileDescriptors in the global DescriptorPool. Otherwise, dependency issues could arise. So if we do not allow runtime-parsed protos to share module space with statically generated "_pb2.py" files (as this and your other comment suggest), then we also cannot allow them to share the default Python descriptor pool.

That being the case, given that underlay pools and fallback DBs are the only mechanisms available to cross-link descriptors from different pools, there is no way to make this code work:

```python
A = protobuf.protos("A.proto")
import B_pb2
```

where, as in my earlier comment, "B.proto" depends on A.proto.

However, you could still make this code work:

```python
import A_pb2
B = protobuf.protos("B.proto")
```

by setting up a separate descriptor pool for runtime-parsed files with the default Python pool as an underlay. If we're disallowing modification of the import machinery, then we could perhaps instantiate `B` above using [`types.ModuleType("B_pb2")`](https://docs.python.org/3/library/types.html#types.ModuleType) (or, equivalently, [`importlib.util.module_from_spec()`](https://docs.python.org/3/library/importlib.html#importlib.util.module_from_spec)). This appears not to introduce clashes:

```python
>>> B = types.ModuleType("B_pb2")
>>> import B_pb2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'B_pb2'
```

Thoughts?